### PR TITLE
Add apply_activation_checkpointing helper function

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #78854
* #78752
* __->__ #78704

Adds `apply_activation_checkpointing` helper function, which will allow users to easily checkpoint multiple submodules within a module depending on some condition. 

The function takes in `model`, `checkpoint_wrap_fn` which indicates the checkpoint_wrapper config, and a lambda indicating whether a module should be wrapped or not. It does this wrapper and updates the model in place, i.e.

`Sequential(Linear, Linear)` would become `Sequential(wrapped(Linear), wrapped(Linear))` in place. 

This is useful for several uses cases we've been seeing that have been trying to apply activation checkpointing, where we've had to hack around the module hierarchy via setattr/getattr which can easily lead to issues if not done properly.

### CODE REVIEW

File: torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py [hunk 1]
```diff
@@ -123,3 +123,54 @@ def checkpoint_wrapper(
...
+            A lambda function which will be passed overall model and current layer and returns
+            ``True`` or ``False`` depending on whether input layer should be wrapped.
+    Returns: None (`model` is modified inplace)
+    """
+    layers = []
+    for name, layer in model.named_modules():
+        if check_fn(model, layer):
+            toks = name.strip().split(".")
+            module = layer
+            layer = model
```
Comment (awgu) [inline torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py:160]: Some interesting variable aliasing going on here 😆 

```suggestion
    for name, module in model.named_modules():
        if check_fn(model, module):
            toks = name.strip().split(".")
            layer = model
```

Comment (awgu) [reply torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py:160]: Overall, this method is a bit hard to follow for me because the variables keep changing what they mean.

It seems like the point is that we have to find `nn.Module` attributes of an `nn.Module` to make sure to set them as attributes of the new `nn.Module` returned by `checkpoint_wrapper_fn()`?

`checkpoint_wrapper_fn()` is only applied to the original `nn.Module`s of `model` (via `new_layer = checkpoint_wrapper_fn(module)` where `module = layer` and `layer` is from `model.named_modules()`).

Why do we only need to `setattr` for `nn.Module` attributes for the new `nn.Module` returned by `checkpoint_wrapper_fn()`? Why are they different from other non-`nn.Module` attributes?

Comment (rohan-varma) [reply torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py:160]: > Why do we only need to setattr for nn.Module attributes for the new nn.Module returned by checkpoint_wrapper_fn()? Why are they different from other non-nn.Module attributes?

We only need to set these since these are new modules, i.e. a `CheckpointWrapper` module containing a submodule which is the wrapped module.

The rest of the modules we didn't wrap are already set? 

Agreed, naming here can definitely be better and function can be made clearer, let me update.

File: torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py [hunk 2]
```diff
@@ -123,3 +123,56 @@ def checkpoint_wrapper(
...
+    """
+    # checkpointed_submodules will create a list of (parent, child_attr, wrapped_child)
+    # which will allow us to set the checkpoint wrapped child on the parent appropriately.
+    checkpointed_submodules = []
+    for name, submodule in model.named_modules():
+        if check_fn(model, submodule):
+            # This submodule should be wrapped with activation checkpointing
+            tokens = name.strip().split(".")
+            submodule_parent = model
+            # Get parent layer of module by iterating through the tokenized FQN.
```
Comment (awgu) [inline torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py:163]: nit: (IIUC)
```suggestion
            # Get parent layer of submodule by iterating through the tokenized FQN.
```

**Review by awgu — COMMENTED** [review]
Thanks for clearing this up. I think it makes sense now.

**Review by zhaojuanmao — COMMENTED** [review]
Wondering whether we can just use FSDP '_recursive_wrap()' API for this helper function as well? it also traverses the model, wrap submodules and change model in place. basically make the wrap_cls=checkpoint_wrapper?

File: torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py [hunk 3]
```diff
@@ -123,3 +123,56 @@ def checkpoint_wrapper(
...
+            for token in tokens[:-1]:
+                # isnumeric() check to handle case where layers are in a list,
+                # see https://discuss.pytorch.org/t/how-to-replace-a-layer-with-own-custom-variant/43586/12
+                if not token.isnumeric():
+                    submodule_parent = getattr(submodule_parent, token)
+                else:
+                    submodule_parent = submodule_parent[int(token)]
```
Comment (zhaojuanmao) [inline torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py:170]: not sure the way to get submodule_parent is robust and general

**Comment by rohan-varma** [comment]
> Wondering whether we can just use FSDP '_recursive_wrap()' API for this helper function as well? it also traverses the model, wrap submodules and change model in place. basically make the wrap_cls=checkpoint_wrapper?

It might require our own policy, as the user might want a more general lambda function? It can be quite similar to transformer_auto_wrap_policy.

File: torch/distributed/fsdp/wrap.py [hunk 4]
```diff
@@ -29,6 +29,43 @@ def always_wrap_policy(*args, **kwargs) -> bool:
...
+           The module to be considered in this decision.
+       recurse (bool):
+           Indicate if this is called to make a decision on whether we
+           should recurse down a subgraph of the module structure.
+           If False, it means this function is called to make a decision
+           on whether we should wrap the said module.
+       unwrapped_params (int):
+           The number of parameters yet to be wrapped in this module.
+
+       transformer_layer_cls (int):
```
Comment (zhaojuanmao) [inline torch/distributed/fsdp/wrap.py:58]: nit: no 'transformer_layer_cls'?

**Review by zhaojuanmao — APPROVED** [review]
some minors, looks clean:)
